nixos/modules/services/misc/matrix-synapse.nix

```diff
@@ -669,30 +669,43 @@ in {
       gid = config.ids.gids.matrix-synapse;
     };
 
-    services.postgresql = mkIf (usePostgresql && cfg.create_local_database) {
-      enable = mkDefault true;
-      ensureDatabases = [ cfg.database_name ];
-      ensureUsers = [{
-        name = cfg.database_user;
-        ensurePermissions = { "DATABASE \"${cfg.database_name}\"" = "ALL PRIVILEGES"; };
-      }];
-    };
+    services.postgresql.enable = mkIf usePostgresql (mkDefault true);
 
     systemd.services.matrix-synapse = {
       description = "Synapse Matrix homeserver";
-      after = [ "network.target" ] ++ lib.optional config.services.postgresql.enable "postgresql.service" ;
+      after = [ "network.target" "postgresql.service" ];
       wantedBy = [ "multi-user.target" ];
       preStart = ''
         ${cfg.package}/bin/homeserver \
           --config-path ${configFile} \
           --keys-directory ${cfg.dataDir} \
           --generate-keys
+      '' + optionalString (usePostgresql && cfg.create_local_database) ''
+        if ! test -e "${cfg.dataDir}/db-created"; then
+          ${pkgs.sudo}/bin/sudo -u ${pg.superUser} \
+            ${pg.package}/bin/createuser \
+            --login \
+            --no-createdb \
+            --no-createrole \
+            --encrypted \
+            ${cfg.database_user}
+          ${pkgs.sudo}/bin/sudo -u ${pg.superUser} \
+            ${pg.package}/bin/createdb \
+            --owner=${cfg.database_user} \
+            --encoding=UTF8 \
+            --lc-collate=C \
+            --lc-ctype=C \
+            --template=template0 \
+            ${cfg.database_name}
+          touch "${cfg.dataDir}/db-created"
```

Comment:
Wouldn't we reintroduce the issues @grahamc expressed with `ensure*`?

Reply:
@Ma27 I don't understand what you mean by "reintroduce" - the issues @grahamc brought up have not been resolved by any approach mentioned so far.

Reply:
This PR is now open since 19 days and stalled because of @grahamc 's opinion on `ensureDatabases`. 

> I would prefer to not see this part of this PR merge.
-- @grahamc 

This is clearly resolved. We cannot guarantee the state of postgresql. State may change outside of Nix's control, whether we like it or not. We will and should not solve this issue here.
Can we please get this into master and backport this to 19.09 already?

@Ma27 I will ping you concerning the SQL stanza.

Reply:
We'll drop the database setup entirely for now. An example is shown in the vm-test, users with existing setups won't have any problems since they already have a fully configured database. For further information, please read the release-notes I added in this PR.